scene/2d/audio_stream_player_2d.cpp

```diff
@@ -33,84 +33,46 @@
+
 	switch (p_what) {
 		case NOTIFICATION_ENTER_TREE: {
 			AudioServer::get_singleton()->add_listener_changed_callback(_listener_changed_cb, this);
-			if (autoplay && !Engine::get_singleton()->is_editor_hint()) {
-				play();
-			}
-			set_stream_paused(!can_process());
 		} break;
 
 		case NOTIFICATION_EXIT_TREE: {
-			set_stream_paused(true);
 			AudioServer::get_singleton()->remove_listener_changed_callback(_listener_changed_cb, this);
 		} break;
 
-		case NOTIFICATION_PREDELETE: {
-			stop();
-		} break;
-
-		case NOTIFICATION_PAUSED: {
-			if (!can_process()) {
-				// Node can't process so we start fading out to silence.
-				set_stream_paused(true);
-			}
-		} break;
-
-		case NOTIFICATION_UNPAUSED: {
-			set_stream_paused(false);
-		} break;
-
 		case NOTIFICATION_INTERNAL_PHYSICS_PROCESS: {
 			// Update anything related to position first, if possible of course.
-			if (setplay.get() > 0 || (active.is_set() && last_mix_count != AudioServer::get_singleton()->get_mix_count()) || force_update_panning) {
+			if (setplay.get() > 0 || (internal.active.is_set() && last_mix_count != AudioServer::get_singleton()->get_mix_count()) || force_update_panning) {
 				force_update_panning = false;
 				_update_panning();
 			}
 
 			if (setplayback.is_valid() && setplay.get() >= 0) {
-				active.set();
-				AudioServer::get_singleton()->start_playback_stream(setplayback, _get_actual_bus(), volume_vector, setplay.get(), pitch_scale);
+				internal.active.set();
+				AudioServer::get_singleton()->start_playback_stream(setplayback, _get_actual_bus(), volume_vector, setplay.get(), internal.pitch_scale);
 				setplayback.unref();
 				setplay.set(-1);
 			}
 
-			if (!stream_playbacks.is_empty() && active.is_set()) {
-				// Stop playing if no longer active.
-				Vector<Ref<AudioStreamPlayback>> playbacks_to_remove;
-				for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
-					if (playback.is_valid() && !AudioServer::get_singleton()->is_playback_active(playback) && !AudioServer::get_singleton()->is_playback_paused(playback)) {
-						playbacks_to_remove.push_back(playback);
-					}
-				}
-				// Now go through and remove playbacks that have finished. Removing elements from a Vector in a range based for is asking for trouble.
-				for (Ref<AudioStreamPlayback> &playback : playbacks_to_remove) {
-					stream_playbacks.erase(playback);
-				}
-				if (!playbacks_to_remove.is_empty() && stream_playbacks.is_empty()) {
-					// This node is no longer actively playing audio.
-					active.clear();
-					set_physics_process_internal(false);
-				}
-				if (!playbacks_to_remove.is_empty()) {
-					emit_signal(SNAME("finished"));
-				}
-			}
-
-			while (stream_playbacks.size() > max_polyphony) {
-				AudioServer::get_singleton()->stop_playback_stream(stream_playbacks[0]);
-				stream_playbacks.remove_at(0);
+			if (!internal.stream_playbacks.is_empty() && internal.active.is_set()) {
+				internal.process();
 			}
+			internal.ensure_playback_limit();
```

Comment:
The playback limit is applied each frame, unlike in AudioStreamPlayer, which does it only on `play()`. Maybe other players could do that too?